File: app/src/main/java/com/example/purepawapp/ui/admin/AdminBookingAdapter.java
package com.example.purepawapp.ui.admin;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.example.purepawapp.data.model.Booking;
import com.example.purepawapp.databinding.ItemAdminBookingBinding;
import com.example.purepawapp.util.CurrencyUtils;

import java.util.function.Consumer;

public class AdminBookingAdapter extends ListAdapter<Booking, AdminBookingAdapter.ViewHolder> {

    private final Consumer<Booking> onClick;

    private static final DiffUtil.ItemCallback<Booking> DIFF_CALLBACK = new DiffUtil.ItemCallback<>() {
        @Override
        public boolean areItemsTheSame(@NonNull Booking oldItem, @NonNull Booking newItem) {
            return oldItem.getId().equals(newItem.getId());
        }

        @Override
        public boolean areContentsTheSame(@NonNull Booking oldItem, @NonNull Booking newItem) {
            return oldItem.getStatus().equals(newItem.getStatus());
        }
    };

    public AdminBookingAdapter(Consumer<Booking> onClick) {
        super(DIFF_CALLBACK);
        this.onClick = onClick;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemAdminBookingBinding binding = ItemAdminBookingBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.bind(getItem(position));
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        private final ItemAdminBookingBinding binding;

        ViewHolder(ItemAdminBookingBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        void bind(Booking booking) {
            binding.tvServiceName.setText(booking.getServiceName());
            binding.tvPetAndDate.setText(booking.getPet().getName() + " · " + booking.getBookingDate() + " " + booking.getTimeSlot());
            binding.tvPrice.setText(CurrencyUtils.toVndString(booking.getPrice()));
            StatusStyle style = AdminStatusUi.bookingStatusStyle(booking.getStatus());
            binding.tvStatus.setText(style.getLabel());
            binding.tvStatus.setTextColor(binding.getRoot().getResources().getColor(style.getTextColorRes(), null));
            binding.tvStatus.setBackgroundTintList(binding.getRoot().getResources().getColorStateList(style.getBgColorRes(), null));
            binding.getRoot().setOnClickListener(v -> onClick.accept(booking));
        }
    }
}

File: app/src/main/java/com/example/purepawapp/ui/admin/AdminBookingAdapter.kt
package com.example.purepawapp.ui.admin

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.example.purepawapp.data.model.Booking
import com.example.purepawapp.databinding.ItemAdminBookingBinding
import com.example.purepawapp.util.toVndString

class AdminBookingAdapter(
    private val onClick: (Booking) -> Unit
) : ListAdapter<Booking, AdminBookingAdapter.ViewHolder>(DIFF_CALLBACK) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = ItemAdminBookingBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    inner class ViewHolder(private val binding: ItemAdminBookingBinding) : RecyclerView.ViewHolder(binding.root) {
        fun bind(booking: Booking) {
            binding.tvServiceName.text = booking.serviceName
            binding.tvPetAndDate.text = "${booking.pet.name} · ${booking.bookingDate} ${booking.timeSlot}"
            binding.tvPrice.text = booking.price.toVndString()
            val style = bookingStatusStyle(booking.status)
            binding.tvStatus.text = style.label
            binding.tvStatus.setTextColor(binding.root.resources.getColor(style.textColorRes, null))
            binding.tvStatus.backgroundTintList =
                binding.root.resources.getColorStateList(style.bgColorRes, null)
            binding.root.setOnClickListener { onClick(booking) }
        }
    }

    companion object {
        private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<Booking>() {
            override fun areItemsTheSame(oldItem: Booking, newItem: Booking): Boolean = oldItem.id == newItem.id
            override fun areContentsTheSame(oldItem: Booking, newItem: Booking): Boolean = oldItem == newItem
        }
    }
}
